src/gretel_trainer/benchmark/session.py

```diff
@@ -0,0 +1,346 @@
+            "Generate time (sec)": generate_time,
+            "Total time (sec)": total_time,
+        }
+
+    def _setup_gretel_run(
+        self,
+        job: JobSpec,
+        artifact_key: Optional[str],
+        trainer_project_index: int,
+    ) -> None:
+        if not isinstance(job.model, GretelModel):
+            raise BenchmarkException(
+                "_setup_gretel_run can only be used with GretelModel!"
+            )
+
+        run_key = job.make_run_key()
+        run_identifier = run_key.identifier
+        strategy = self._set_gretel_strategy(
+            job=job,
+            run_identifier=run_identifier,
+            trainer_project_index=trainer_project_index,
+            artifact_key=artifact_key,
+        )
+        executor = Executor(
+            strategy=strategy,
+            run_identifier=run_identifier,
+            evaluate_project=self._project,
+            config=self._config,
+        )
+        self._gretel_executors[run_key] = executor
+
+    def _setup_custom_run(
+        self,
+        job: JobSpec,
+        artifact_key: Optional[str] = None,
+    ) -> None:
+        if isinstance(job.model, GretelModel):
+            raise BenchmarkException(
+                "_setup_custom_run cannot be used with GretelModel!"
+            )
+
+        run_key = job.make_run_key()
+        run_identifier = run_key.identifier
+        strategy = CustomStrategy(
+            benchmark_model=job.model,
+            dataset=job.dataset,
+            run_identifier=run_identifier,
+            config=self._config,
+            artifact_key=artifact_key,
+        )
+        executor = Executor(
+            strategy=strategy,
+            run_identifier=run_identifier,
+            evaluate_project=self._project,
+            config=self._config,
+        )
+        self._custom_executors[run_key] = executor
+
+    def _basic_status(self, future: Future) -> str:
+        if future.done():
+            return "Finished"
+        elif future.cancelled():
+            return "Cancelled"
+        else:
+            return "Running"
+
+    # TODO(pm): thoughts on renaming this to e.g. "_create_gretel_strategy()"?
+    #  From the name, I'd expect something to be set on the `self`, but it's returned,
+    #  so it's a bit confusing. Unless "set" means something else here.
+    def _set_gretel_strategy(
+        self,
+        job: JobSpec,
+        run_identifier: str,
+        trainer_project_index: int,
+        artifact_key: Optional[str],
+    ) -> Union[GretelSDKStrategy, GretelTrainerStrategy]:
+        # TODO(pm): this is to make typechecker happy, couldn't figure out a nicer way
+        #  to do this, without much refactoring.
+        #  Why I want to have job here? Because I needed to pass additional context
+        if not isinstance(job.model, GretelModel):
+            raise BenchmarkException(
+                "_setup_gretel_run can only be used with GretelModel!"
+            )
```

Comment:
I think we would have to make `JobSpec` generic with respect to model type, and then for example constrain this method to take a `JobSpec[GretelModel]`. Been awhile since I've played with generics in python typehints, not sure how much effort would be required / if its worth it 🤷

Reply:
Good idea, I was thinking about that, but didn't know if that would be a path worth taking. With your comment I went into the rabbit whole and this is what I got to, which works relatively well:
- it fails type check in places where I'd expect it to and requires `cast` (it cannot figure things out on its own, as `isinstance` cannot be used with generic type param)
- is reasonably easy to reason about/read/write

```python
AnyModelType = Union[GretelModel, CustomModel]
MODEL_BASE_TYPE = TypeVar('MODEL_BASE_TYPE', bound=AnyModelType, covariant=True)
# covariant=True, so that `list[GretelModel]` can be used where `list[AnyModelType]` is used

@dataclass
class JobSpec(Generic[MODEL_BASE_TYPE]):
    dataset: Dataset
    model: MODEL_BASE_TYPE
    
    def make_run_key(self):
        return RunKey((model_name(self.model), self.dataset.name))


class RandomModel(CustomModel):
    def train(self, source: Dataset, **kwargs) -> None:
        ...

    def generate(self, **kwargs) -> pd.DataFrame:
        ...

def foo(job: JobSpec[GretelModel]):
    print(job.model.name)

def bar(jobs: list[JobSpec[AnyModelType]]):
    for job in jobs:
		foo(job) # fails type check
        if isinstance(job.model, GretelModel):
			foo(job) # fails type check
            foo(cast(JobSpec[GretelModel], job)) # ok
			# note: that `cast` is needed here, as there is no way to "isinstance" based on the generic type, and the `isinstance` on `job.model` doesn't translate to job being `JobSpec[GretelModel]`.

if __name__ == '__main__':
    job = JobSpec(Dataset("a"), RandomModel())
    bar([job]) # ok, list[JobSpec[RandomModel]] can be used where list[JobSpec[AnyModelType]] is required, because of `covariant=True`
    foo(job) # fails type check
```

What do you think? :)

Reply:
Very cool, I didn't know about `TypeGuard`, but it looks like exactly what's needed to solve this. Good to know we could use it with `typing_extensions`, even with our Python version.

Evaluating both options - `cast` vs `TypeGuard`, since the usage is quite limited, we'd needed it in a single spot right now, I'm wondering if adding extra dependency to have `TypeGuard` is worth in this case. 
My initial thinking is to just keep the `cast`, but since it's a small library without any dependencies, we could also just do that. Are there any other spots it could be useful?

Reply:
1. Potentially could be useful [here](https://github.com/gretelai/trainer/blob/b2/src/gretel_trainer/benchmark/comparison.py#L94-L101)? The current code block isn't that bad, but I'd be curious if TypeGuard could clean it up a little.
2. A while back I [pinned pyright](https://github.com/gretelai/trainer/pull/105) when CI unexpectedly pulled a new release that caught many more type warnings that I didn't have time to address. I remember one of the main categories of warnings had to do with the difference between `Job` and `Model | RecordHandler`. I'm not certain of this, but I suspect we might want to ditch using the client `Job` class in our type hints and replace it with a generic that gets narrowed as needed via TypeGuard to either of the two subclasses (model/RH).

🤷

Reply:
Thanks! Added `typing-extensions` and using `TypeGuard` in that spot now.